### PR TITLE
Debug information for immutable references.

### PR DESCRIPTION
**Pull request by ekpyron**
Reminder: needs documentation when format is decided.

### CODE REVIEW

File: test/cmdlineTests/standard_immutable_references/output.json [hunk 1]
```diff
@@ -0,0 +1,2 @@
+{"contracts":{"a.sol":{"A":{"evm":{"deployedBytecode":{"immutableReferences":{"a.sol:A.x (3)":[77]},"linkReferences":{},"object":"bytecode removed","opcodes":"opcodes removed","sourceMap":"0:96:0:-:0;;;;5:9:-1;2:2;;;27:1;24;17:12;2:2;0:96:0;;;;;;;;;;;;;;;;12:1:-1;9;2:12;38:56:0;;;:::i;:::-;;;;;;;;;;;;;;;;;;;;72:7;90:1;83:8;;38:56;:::o"}}}}},"errors":[{"component":"general","formattedMessage":"a.sol: Warning: Source file does not specify required compiler version!
```
Comment (ekpyron) [inline test/cmdlineTests/standard_immutable_references/output.json]: ``"a.sol:A.x (3)":[77]``
I'm not sure... maybe AST IDs only would be better... or just plain identifiers ``x``... but it'll be weirder with multi-slot types - I'll push another example with function pointers....

File: test/cmdlineTests/standard_immutable_references_function_pointer/output.json [hunk 2]
```diff
@@ -0,0 +1,2 @@
+{"contracts":{"a.sol":{"A":{"evm":{"deployedBytecode":{"immutableReferences":{"a.sol:A.x (10) address":[73],"a.sol:A.x (10) functionIdentifier":[106]},"linkReferences":{},"object":"bytecode removed","opcodes":"opcodes removed","sourceMap":"0:110:0:-:0;;;;5:9:-1;2:2;;;27:1;24;17:12;2:2;0:110:0;;;;;;;;;;;;;;;;;;;;;12:1:-1;9;2:12;80:28:0;;;:::i;:::-;;13:24;;;:::i;:::-;;80:28;102:1;;:3;;;;;;;;;;;;;;;;;;;;;;;5:9:-1;2:2;;;27:1;24;17:12;2:2;102:3:0;;;;8:9:-1;5:2;;;45:16;42:1;39;24:38;77:16;74:1;67:27;5:2;102:3:0;;;;80:28::o;13:24::-;:::o"}}}}},"errors":[{"component":"general","formattedMessage":"a.sol: Warning: Source file does not specify required compiler version!
```
Comment (ekpyron) [inline test/cmdlineTests/standard_immutable_references_function_pointer/output.json]: Yeah - here the stack slot name is also appended to it...

Comment (chriseth) [reply test/cmdlineTests/standard_immutable_references_function_pointer/output.json]: actually, the value `73` will change with changes to the code generator :)

File: libsolidity/interface/StandardCompiler.cpp [hunk 3]
```diff
@@ -309,13 +310,30 @@ Json::Value formatLinkReferences(std::map<size_t, std::string> const& linkRefere
...
 
-Json::Value collectEVMObject(evmasm::LinkerObject const& _object, string const* _sourceMap)
+Json::Value formatImmutableReferences(std::map<std::string, std::vector<size_t>> const& _immutableReferences)
+{
+	Json::Value ret(Json::objectValue);
+
+	for (auto const& [name, occurrences]: _immutableReferences)
+	{
+		Json::Value array(Json::arrayValue);
+		for (auto occurrence: occurrences)
```
Comment (chriseth) [inline libsolidity/interface/StandardCompiler.cpp]:
```suggestion
		for (size_t occurrence: occurrences)
```

File: libsolidity/interface/StandardCompiler.cpp [hunk 4]
```diff
@@ -309,13 +310,30 @@ Json::Value formatLinkReferences(std::map<size_t, std::string> const& linkRefere
 	return ret;
 }
 
-Json::Value collectEVMObject(evmasm::LinkerObject const& _object, string const* _sourceMap)
+Json::Value formatImmutableReferences(std::map<std::string, std::vector<size_t>> const& _immutableReferences)
+{
+	Json::Value ret(Json::objectValue);
+
+	for (auto const& [name, occurrences]: _immutableReferences)
```
Comment (chriseth) [inline libsolidity/interface/StandardCompiler.cpp]:
```suggestion
	for (auto const& [name, bytecodeOffset]: _immutableReferences)
```

File: libevmasm/LinkerObject.h [hunk 5]
```diff
@@ -44,6 +44,10 @@ struct LinkerObject
 	/// that refer to their values.
 	std::map<u256, std::vector<size_t>> immutableReferences;
 
+	/// Map from immutable variable identifiers to a list of offsets into the runtime bytecode used to produce debug
+	/// information.
+	std::map<std::string, std::vector<size_t>> immutableDebugInfo;
```
Comment (chriseth) [inline libevmasm/LinkerObject.h]: I think the use goes far beyond debug information...

File: libevmasm/LinkerObject.h [hunk 6]
```diff
@@ -44,6 +44,10 @@ struct LinkerObject
 	/// that refer to their values.
 	std::map<u256, std::vector<size_t>> immutableReferences;
```
Comment (chriseth) [inline libevmasm/LinkerObject.h]: Does it make sense to use only one mapping for both purposes?

File: libevmasm/Assembly.cpp [hunk 7]
```diff
@@ -543,6 +543,9 @@ LinkerObject const& Assembly::assemble() const
 			);
 			immutableReferencesBySub = linkerObject.immutableReferences;
 		}
+		for (auto const& [immutable, values]: linkerObject.immutableDebugInfo)
+			for (auto value: values)
+				ret.immutableDebugInfo[immutable].emplace_back(value);
```
Comment (chriseth) [inline libevmasm/Assembly.cpp]: Should we really copy it into the creation time linker object?

**Comment by chriseth** [comment]
Please don't forget the metadata!

**Comment by ekpyron** [comment]
We don't have link references in the metadata yet either... and I'm not sure there's a natural place for either of them so far... so we should probably first decide where to put both of those exactly? Just besides the abi? Also: so far the metadata doesn't depend on code generation, so that would also change...

**Comment by chriseth** [comment]
But I think it would be really useful to mark places in the (runtime) bytecode that could be different from the ones generated by recompiling. On the other hand, if you recompile, you get them anyway....

**Comment by ekpyron** [comment]
Independently of the metadata question, I'm not quite happy with the identifiers as keys... they actually need to be parsed the way they are... or generated from the AST... plus we fix the names of the stack slot identifiers...

I'm wondering whether it'd be nicer to have:
```
  astId: [ [offsetOfStackSlot1OfOccurrence1, offsetOfStackSlot2OfOccurrence1], [offsetOfStackSlot1OfOccurrence2, offsetOfStackSlot2OfOccurrence2]]
```
That would be more effort to generate from what we store now, but maybe it'd be worth it...

**Comment by chriseth** [comment]
I don't think we should provide fine-grained stack slot resolution, but probably lengths would make sense. This would also allow use to switch to a codecopy-based method later.

**Comment by ekpyron** [comment]
So I just include the contained PUSH32 opcodes in the bytecode range for multi-slot immutables?

**Comment by chriseth** [comment]
Ah, forgot about the push32 opcodes...

**Comment by chriseth** [comment]
So to make it proper we would need something like:
```
[// for each reference
  [// for each "part"
    [offset, size]
  ]
]
```

**Comment by ekpyron** [comment]
We should probably also consider that we may want to switch from push32 to push[justBigEnough].

**Comment by ekpyron** [comment]
> So to make it proper we would need something like:
> 
> ```
> [// for each reference
>   [// for each "part"
>     [offset, size]
>   ]
> ]
> ```

Yes exactly... but it's not really nice either :-).

File: test/cmdlineTests/standard_immutable_references_function_pointer/output.json (hunk 2 above)
Comment (ekpyron) [reply test/cmdlineTests/standard_immutable_references_function_pointer/output.json]: Hopefully that won't happen all the time :-).

**Comment by ekpyron** [comment]
I now switched to offset/size for the current identifiers...
Which is better? Somewhat weird fragile string identifiers or cryptic ID-indexed deeply nested arrays? I'm really not sure :-)... maybe there's still an entirely different and better way...

**Comment by ekpyron** [comment]
A potential advantage of the strings is that they are the same ones that occur e.g. in json assembly...
We could also just change those strings - the AST ID plus a stack slot name (if any) should be enough...

**Comment by ekpyron** [comment]
Ok I changed the identifiers now to ``<ASTID> <spaceSeparatedListOfStackSlotNames>``, so e.g.
``10 address`` and ``10 functionIdentifier`` for a function pointer or just ``12`` for an integer.

Instead we could also go for ``10.0`` and ``10.1``, i.e. just use numeric IDs for the stack slots instead of names...

**Comment by ekpyron** [comment]
Or... we could just disallow external function types for the release and sort this out later :-). They are the only value types with more than one stack slot...

**Comment by ekpyron** [comment]
Especially since we could also still decide to pack the external function pointers into a single slot just as in memory...

**Comment by ekpyron** [comment]
I added documentation and while doing so I actually changed this to now:
```
"immutableReferences": {
 "<astId>": [ { "start": 42, "length": 32 }]
}
```
That's how link references look and this way we can add more fields later if needed (e.g. an "offset" might be useful for an array slice using a codecopy approach, etc.).

**Review by chriseth — APPROVED** [review]

**Pull request merged**